solc/CommandLineInterface.cpp

```diff
@@ -687,6 +727,23 @@ void CommandLineInterface::printLicense()
 	sout() << licenseText << endl;
 }
 
+void CommandLineInterface::assembleFromEvmAssemblyJson()
+{
+	solAssert(m_options.input.mode == InputMode::EVMAssemblerJSON);
+	try
+	{
+		solAssert(m_fileReader.sourceUnits().size() == 1);
+		auto const iter = m_fileReader.sourceUnits().begin();
+		m_compiler = make_unique<CompilerStack>(m_universalCallback.callback());
+		m_compiler->importFromEVMAssemblyStack(iter->first, iter->second);
```

Comment:
> The point of having EVMAssemblyStack was to avoid having to use the whole CompilerStack when doing assembly import.

Ok, you're right. I just tried to reuse as much code as possible from the `CompilerStack` - but of course there are always a trade-offs. However, so what I will basically do now is to move the stuff to `EVMAssemblyStack` where I will probably need to duplicate some code from `CompilerStack`. I will then be able to use `EVMAssemblyStack` directly inside `CommandLineInterface`. So there is no need to use the `CompilerStack` at all for the import. 

@cameel Did you mean this? Did I understand you correctly?

Reply:
Probably. Can you say what specifically you need to duplicate? Because I far as I can tell, you may need some small methods like accessors for the bytecode and all, or maybe something to invoke JSON export, but nothing large should be duplicated. If there's something larger that I'm overlooking, then maybe `CompilerStack` does need to keep (or locally create) an instance of `EVMAssemblyStack` after all.

Reply:
After looking a bit closer, I really think it's better to do the import with the support of `CompilerStack`. E.g. if you look at `CommandLineInterface ::handleCombinedJSON` it is right now using the `CompilerStack` directly, in the context of the evm asm import it will use it to retrieve `bin`, `bin-runtime`, `opcodes`, `asm` that might be easy to movable (we will actually need to duplicate the code) to `EVMAssemblyStack`, however we will also need to have a switch for that inside `CommandLineInterface ::handleCombinedJSON`(evm assembly import vs. normal compilation). The other things where related to source-mappings, e.g. `srcmap` and `srcmap-runtime` - those mappings are mainly computed by `evmasm::AssemblyItem::computeSourceMapping` - those may be also easily movable (to be more clear: copyable - copy of `CompilerStack` code into `EVMAssemblyStack` ). We would need to add special code to `handleCombinedJSON`, `handleBytecode` and `handleEVMAssembly` to support normal compilation and the import case. Or we just write a new import specific handler function, that will just use the `EVMAssemblyStack`

Hmm.. I'm really not sure what is better.. right now we don't have code duplication and the `CompilerStack` can be used to import assembly json.. if we change that, we will need to write special handler functions, that will work for the import case and use `EVMAssemblyStack`..  also we will have duplicated code in `EVMAssemblyStack`. Therefore, in the worst case, it could also take some time to implement the needed functions in the `EVMAssemblyStack` correctly. Thats why I was preferring to just reuse the code already present in the `CompilerStack` - because we know it's already working correct.

But maybe in the end it is not that complicated to do this, but we will have more code in the end (with code duplication). I'm not sure what is the better solution. What is your opinion now?

Reply:
However, I think the best is just to try it and then lets see.

Reply:
It's really not that complicated. I gave it a try just now and I managed to do it. Check out [`import-asm-json-updated-with-abstract-assembly-stack`](https://github.com/ethereum/solidity/tree/import-asm-json-updated-with-abstract-assembly-stack). If you think the code is fine you can pull the commits from it into your branch and delete it (and modify those commits however you want or even squash them in to the ones you have now).

I wouldn't worry too much about potential duplication in `CommandLineInterface` because it's slated to be rewritten from scratch at some point in the future anyway. And especially not in `handleCombinedJSON()` because we'll be dropping `--combined-json` sooner or later. We should rather put more weight on keeping `CompilerStack` sensible.

Nevertheless, I think I managed to reconcile both. I introduced an interface you can use to access both `EVMAssemblyStack` and `CompilerStack`. Now the CLI can easily pull the same data from both.

> The other things where related to source-mappings, e.g. `srcmap` and `srcmap-runtime` - those mappings are mainly computed by `evmasm::AssemblyItem::computeSourceMapping` - those may be also easily movable (to be more clear: copyable - copy of `CompilerStack` code into `EVMAssemblyStack` ).

They're pretty small and even then only a fragment of their code is needed in `EVMAssemblyStack`. And I think that in the end we should be able to get rid of the `yulUtilityFileName()` bit, which would make them even smaller.


BTW, I also pushed another commit with small tweaks directly to your branch.

Reply:
Ah cool, great! I just checked your code briefly - it looks really good. I guess I will just re-use your code here. I also managed to extract the stuff from `CompilerStack`, but your solution is cleaner - because of the introduction of that `AbstractAssemblyStack`.

Reply:
After doing #14171 I think it might have been better to let the EVM assembly import stuff in `CompilerStack`. I will do the standard json import for EVM assembly probably later today. But yeah, instead of just using the `CompilerStack` within `StandardCompiler` I need to use now the `EVMAssemblyStack` there - so we will have here more extra-logic. However, let's see how it will look like - maybe it's not that bad.